aws/resource_aws_waf_rate_based_rule.go

```diff
@@ -0,0 +1,219 @@
+			RateKey:     aws.String(d.Get("rate_key").(string)),
+			RateLimit:   aws.Int64(int64(d.Get("rate_limit").(int))),
+		}
+
+		return conn.CreateRateBasedRule(params)
+	})
+	if err != nil {
+		return err
+	}
+	resp := out.(*waf.CreateRateBasedRuleOutput)
+	d.SetId(*resp.Rule.RuleId)
+	return resourceAwsWafRateBasedRuleUpdate(d, meta)
+}
+
+func resourceAwsWafRateBasedRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	params := &waf.GetRateBasedRuleInput{
+		RuleId: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetRateBasedRule(params)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "WAFNonexistentItemException" {
+			log.Printf("[WARN] WAF Rate Based Rule (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	var predicates []map[string]interface{}
+
+	for _, predicateSet := range resp.Rule.MatchPredicates {
+		predicate := map[string]interface{}{
+			"negated": *predicateSet.Negated,
+			"type":    *predicateSet.Type,
+			"data_id": *predicateSet.DataId,
+		}
+		predicates = append(predicates, predicate)
+	}
+
+	d.Set("predicates", predicates)
+	d.Set("name", resp.Rule.Name)
+	d.Set("metric_name", resp.Rule.MetricName)
+	d.Set("rate_key", resp.Rule.RateKey)
+	d.Set("rate_limit", resp.Rule.RateLimit)
+
+	return nil
+}
+
+func resourceAwsWafRateBasedRuleUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	if d.HasChange("predicates") {
+		o, n := d.GetChange("predicates")
+		oldP, newP := o.(*schema.Set).List(), n.(*schema.Set).List()
+		rateLimit := aws.Int64(int64(d.Get("rate_limit").(int)))
+
+		err := updateWafRateBasedRuleResource(d.Id(), oldP, newP, rateLimit, conn)
+		if err != nil {
+			return fmt.Errorf("Error Updating WAF Rule: %s", err)
+		}
+	}
+
+	return resourceAwsWafRateBasedRuleRead(d, meta)
+}
+
+func resourceAwsWafRateBasedRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	oldPredicates := d.Get("predicates").(*schema.Set).List()
+	if len(oldPredicates) > 0 {
+		noPredicates := []interface{}{}
+		rateLimit := aws.Int64(int64(d.Get("rate_limit").(int)))
+
+		err := updateWafRateBasedRuleResource(d.Id(), oldPredicates, noPredicates, rateLimit, conn)
+		if err != nil {
+			return fmt.Errorf("Error updating WAF Rate Based Rule Predicates: %s", err)
+		}
+	}
+
+	wr := newWafRetryer(conn, "global")
+	_, err := wr.RetryWithToken(func(token *string) (interface{}, error) {
+		req := &waf.DeleteRateBasedRuleInput{
+			ChangeToken: token,
+			RuleId:      aws.String(d.Id()),
+		}
+		log.Printf("[INFO] Deleting WAF Rate Based Rule")
+		return conn.DeleteRateBasedRule(req)
+	})
+	if err != nil {
+		return fmt.Errorf("Error deleting WAF Rate Based Rule: %s", err)
+	}
+
+	return nil
+}
+
+func updateWafRateBasedRuleResource(id string, oldP, newP []interface{}, rateLimit *int64, conn *waf.WAF) error {
```

Comment:
It might be more convenient to accept `rateLimit int` here instead and do all the pointer magic inside the function to reduce the repetition, but no biggie.